src/Utilities/FileSystem.hpp

```diff
@@ -0,0 +1,139 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+/// \file
+/// Declares functions to do file system manipulations
+
+#pragma once
+
+#include <string>
+#include <vector>
+
+/*!
+ * \ingroup FileSystem
+ * \brief A light-weight file system library based on POSIX.
+ *
+ * We use this library instead of a subprocess based library because OpenMPI
+ * does not support forking of processes on all systems. Since the
+ * parallelization library we use may be implemented on top of OpenMPI we
+ * take the safe route and use POSIX.
+ */
+namespace filesystem {
+
+/*!
+ * \ingroup FileSystem
+ * \brief Wraps the dirname function to get the pathname of the parent directory
+ *
+ * See the opengroup documentation:
+ * http://pubs.opengroup.org/onlinepubs/009695399/functions/dirname.html
+ *
+ * \example
+ * \snippet Test_FileSystem.cpp dirname
+ *
+ * \requires `file_path` is a valid path on the filesystem
+ * \returns the path to the file
+ */
+std::string dirname(const std::string& path);
+
+/*!
+ * \ingroup FileSystem
+ * \brief Given a path to a file returns the file name
+ *
+ * \example
+ * \snippet Test_FileSystem.cpp get_file_name
+ *
+ * \requires `file_path` is a valid path on the filesystem
+ * \returns the file name
+ */
+std::string get_file_name(const std::string& file_path);
+
+/*!
+ * \ingroup FileSystem
+ * \brief Get the absolute path, resolving symlinks
+ *
+ * \requires `rel_path` is a valid path on the filesystem
+ * \returns the absolute path
+ */
+std::string get_absolute_path(const std::string& rel_path);
+
+/*!
+ * \ingroup FileSystem
+ * \brief Creates a directory, including any parents that don't exist. If the
+ * directory exists `create_directory` does nothing.
+ *
+ * \requires permissions to create `dir` on the filesystem
+ * \effects creates the directory `dir` on the filesystem
+ *
+ * \param dir the path where to create the directory
+ * \param wait_time time to wait between failures
+ * \param num_tries number of attempts to create directory (for slow
+ * filesystems)
+ */
+void create_directory(const std::string& dir, double wait_time = 1,
+                      size_t num_tries = 40);
+
+/*!
+ * \ingroup FileSystem
+ * \brief Returns true if the directory exists
+ *
+ * \returns `true` if the directory exists
+ */
+bool check_if_dir_exists(const std::string& dir);
+
+/*!
+ * \ingroup FileSystem
+ * \brief Returns true if the regular file or link to the regular file exists.
+ *
+ * \note See the stat(2) documentation, e.g. at
+ * http://man7.org/linux/man-pages/man2/stat.2.html for details.
+ *
+ * \returns `true` if the directory exists
+ */
+bool check_if_file_exists(const std::string& file);
+
+/*!
+ * \ingroup FileSystem
+ * \brief Returns true if the path points to a regular file or a link to a
+ * regular file.
+ *
+ * \note See the stat(2) documentation, e.g. at
+ * http://man7.org/linux/man-pages/man2/stat.2.html for details.
+ *
+ * \requires `file` is a valid path on the filesystem
```

Comment:
shouldn't the parameter be named path instead of file as we are determining whether it is a file?